File: python/2616.py
# Link: https://leetcode.com/problems/minimize-the-maximum-difference-of-pairs/description/
# Runtime: 1010 ms
# Memory: 30.89 MB

class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        sorted_nums = nums[:]
        sorted_nums.sort()

        left, right = 0, sorted_nums[-1] - sorted_nums[0]
        while left <= right:
            middle = left + (right - left) // 2

            if self.__get_number_of_pairs(sorted_nums, middle) >= p: right = middle - 1
            else: left = middle + 1
        
        return left
        
    def __get_number_of_pairs(self, sorted_nums: List[int], difference: int) -> int:
        i, number_of_pairs = 0, 0
        while (i < len(sorted_nums) - 1):
            if sorted_nums[i + 1] - sorted_nums[i] <= difference:
                number_of_pairs += 1
                i += 1
            
            i += 1
        
        return number_of_pairs

File: c++/2616.cpp
// Link: https://leetcode.com/problems/minimize-the-maximum-difference-of-pairs/description/
// Runtime: 106 ms
// Memory: 83.53 MB

class Solution {
public:
  int minimizeMax(const std::vector<int> &nums, int p) {
    std::vector<int> sortedNums = nums;
    std::sort(sortedNums.begin(), sortedNums.end());

    int left = 0, right = sortedNums[sortedNums.size() - 1] - sortedNums[0];
    while (left <= right) {
      int middle = left + (right - left) / 2;

      if (getNumberOfPairs(sortedNums, middle) >= p)
        right = middle - 1;
      else
        left = middle + 1;
    }

    return left;
  }

private:
  int getNumberOfPairs(const std::vector<int> &sortedNums, int difference) {
    int i = 0, numberOfPairs = 0;
    while (i < static_cast<int>(sortedNums.size()) - 1) {
      if (sortedNums[i + 1] - sortedNums[i] <= difference) {
        numberOfPairs++;
        i++;
      }

      i++;
    }

    return numberOfPairs;
  }
};
